src/backend_telemetry_buffer.rs

```diff
@@ -0,0 +1,268 @@
+    // without touching Rust code.
+    extension_sql!(
+        r#"
+        CREATE TYPE _prom_ext.backend_telemetry_rec AS (
+            time timestamp with time zone,
+            value BIGINT,
+            tags text[]
+        );
+        "#,
+        name = "backend_telemetry_rec_decl"
+    );
+
+    const COMPOSITE_TYPE_NAME: &str = "_prom_ext.backend_telemetry_rec";
+    const BUFFER_SIZE: usize = 10000;
+
+    type BufferItem = composite_type!(COMPOSITE_TYPE_NAME);
+
+    /// Holds a PG memory context and a ring buffer of pointer- [`Datum`]
+    /// representing [`BufferItem`] inside the memory context.
+    ///
+    /// The ring buffer can hold up to [`BUFFER_SIZE`] items.
+    ///
+    /// The memory context is created as a child of
+    /// [`PgMemoryContexts::CacheMemoryContext`]
+    struct Buffer {
+        mem_ctx: PgMemoryContexts,
+        inner_buffer: PgBox<[Datum; BUFFER_SIZE], AllocatedByRust>,
+        next_idx: usize,
+    }
+
+    /// A helper, for iterating over the [`Buffer`].
+    /// Clears the buffer when dropped.
+    struct BufferIter<'b> {
+        buffer: &'b mut Buffer,
+        pos: usize,
+    }
+
+    impl Buffer {
+        fn new() -> Self {
+            let mem_ctx = PgMemoryContexts::CacheMemoryContext
+                .switch_to(|_| PgMemoryContexts::new("backend_telemetry_buffer_context"));
+            let inner_buffer = PgBox::<[Datum; BUFFER_SIZE], AllocatedByRust>::alloc0_in_context(
+                PgMemoryContexts::CacheMemoryContext,
+            );
+            Self {
+                mem_ctx,
+                inner_buffer,
+                next_idx: 0,
+            }
+        }
+
+        /// Copies [`BufferItem`] into the memory context associated
+        /// with the [`Buffer`]. Frees an older item if it needs
+        /// to be overwritten due to ring buffer overflow.
+        fn append(&mut self, item: BufferItem) -> bool {
+            let copied_opt = self.mem_ctx.switch_to(|_| item.into_composite_datum());
+            let fully_initialized = self.next_idx >= BUFFER_SIZE;
+
+            self.inner_buffer
+                .get_mut(self.next_idx % BUFFER_SIZE)
+                .and_then(|place| {
+                    copied_opt.map(|copied| {
+                        if fully_initialized {
+                            unsafe { pg_sys::pfree(place.cast_mut_ptr()) }
+                        }
+                        *place = copied;
+                    })
+                })
+                .map(|_| self.next_idx += 1)
+                .is_some()
+        }
+
+        /// Empties the buffer and frees memory occupied by contained items.
+        fn reset(&mut self) {
+            self.next_idx = 0;
+            self.mem_ctx.reset();
+        }
+
+        /// Creates an iterator, traversing all initialized items
+        /// in the buffer. Calls `reset()` when dropped.
+        fn consume_as_iter(&mut self) -> BufferIter {
```

Comment:
You could replace this method with the following for a little more 🦀 -idiomism:

```rust
    impl<'a> IntoIterator for &'a mut Buffer {
        type Item = BufferItem;
        type IntoIter = BufferIter<'a>;

        #[inline]
        fn into_iter(self) -> BufferIter<'a> {
            BufferIter {
                buffer: self,
                pos: 0,
            }
        }
    }
```

Reply:
Thanks! Will do!